lib/elixir/lib/calendar/date.ex

```diff
@@ -409,15 +409,24 @@ defmodule Date do
   @spec to_iso8601(Calendar.date(), :extended | :basic) :: String.t()
   def to_iso8601(date, format \\ :extended)
 
-  def to_iso8601(%{calendar: Calendar.ISO} = date, format) when format in [:basic, :extended] do
+  def to_iso8601(date, format) do
+    to_iso8601_iodata(date, format)
+    |> IO.iodata_to_binary()
+  end
+
+  @spec to_iso8601_iodata(Calendar.date(), :extended | :basic) :: iodata
+  def to_iso8601_iodata(date, format \\ :extended)
+
+  def to_iso8601_iodata(%{calendar: Calendar.ISO} = date, format)
```

Comment:
Let’s not expose these functions from the date/time/naivedatetime/datetime APIs yet, only use it internally.